test/config-luatest/credentials_applier_test.lua

```diff
@@ -783,3 +783,255 @@ g.test_sync_ro_rw = function(g)
+                        },
+                    },
+                },
+            },
+        },
+    }
+
+    helpers.reload_success_case(g, {
+        options = iconfig,
+        verify = function()
+            box.schema.space.create('myspace1')
+            box.schema.func.create('myfunc2')
+            box.schema.sequence.create('myseq3')
+
+            local internal =
+                    require('internal.config.applier.credentials')._internal
+
+            local perm_1 = box.schema.user.info('myuser1')
+            perm_1 = internal.privileges_from_box(perm_1)
+
+            local perm_2 = box.schema.user.info('myuser2')
+            perm_2 = internal.privileges_from_box(perm_2)
+
+            t.assert_equals(perm_1['space']['myspace1'],
+                                {read = true, write = true})
+            t.assert_equals(perm_2['space']['myspace1'],
+                                {read = true, write = true})
+            t.assert_equals(perm_1['function']['myfunc2'],
+                                {execute = true})
+            t.assert_equals(perm_2['function']['myfunc2'], nil)
+            t.assert_equals(perm_1['sequence']['myseq3'], nil)
+            t.assert_equals(perm_2['sequence']['myseq3'],
+                                {read = true, write = true})
+
+            t.assert_equals(perm_1['space']['myspace2'], nil)
+            t.assert_equals(perm_2['space']['myspace3'], nil)
+            t.assert_equals(perm_1['function']['myfunc1'], nil)
+            t.assert_equals(perm_2['function']['myfunc3'], nil)
+            t.assert_equals(perm_1['sequence']['myseq1'], nil)
+            t.assert_equals(perm_2['sequence']['myseq1'], nil)
+        end,
+        options_2 = iconfig,
+        verify_2 = function(iconfig)
+            box.schema.space.create('myspace2')
+            box.schema.space.create('myspace3')
+            box.schema.func.create('myfunc1')
+            box.schema.func.create('myfunc3')
+            box.schema.sequence.create('myseq1')
+            box.schema.sequence.create('myseq2')
+
+            local internal =
+                    require('internal.config.applier.credentials')._internal
+
+            local perm_1 = box.schema.user.info('myuser1')
+            perm_1 = internal.privileges_from_box(perm_1)
+
+            local perm_2 = box.schema.user.info('myuser2')
+            perm_2 = internal.privileges_from_box(perm_2)
+
+            local exp_perm_1 = iconfig.credentials.users.myuser1
+            exp_perm_1 = internal.privileges_from_config(exp_perm_1)
+            exp_perm_1 = internal.privileges_add_defaults('myuser1', 'user',
+                                                          exp_perm_1)
+
+            local exp_perm_2 = iconfig.credentials.users.myuser2
+            exp_perm_2 = internal.privileges_from_config(exp_perm_2)
+            exp_perm_2 = internal.privileges_add_defaults('myuser2', 'user',
+                                                          exp_perm_2)
```

Comment:
Here we validate tarantool's behavior that depends on the given `privileges_*` functions using these functions. Comparing to a simple-but-large test with hardcoded expected values this approach may miss some mistakes.

Let's assume that `privileges_from_config` should return some privilege, but it doesn't. It means that tarantool will not grant the given privilege. However, the test will also think that the privilege shouldn't be granted.

I'm in-doubt-but-ok regarding using `privileges_from_box` in test oracles. However, using `privileges_from_config` and `privileges_from_defaults` seems dangerous for me.

Reply:
Ok, will do.

Reply:
Added additional manual checks for both tests.